src/Editor/ImGuizmo/imguizmo.cpp

```diff
@@ -0,0 +1,1719 @@
+            }
+
+            // snap
+            if (snap)
+            {
+                float scaleSnap[] = { snap[0], snap[0], snap[0] };
+                ComputeSnap(gContext.mScale, scaleSnap);
+            }
+
+            // no 0 allowed
+            for (int i = 0; i < 3; i++)
+                gContext.mScale[i] = max(gContext.mScale[i], 0.001f);
+
+            // compute matrix & delta
+            matrix_t deltaMatrixScale;
+            deltaMatrixScale.Scale(gContext.mScale * gContext.mScaleValueOrigin);
+
+            matrix_t res = deltaMatrixScale * gContext.mModel;
+            *(matrix_t*)matrix = res;
+
+            if (deltaMatrix)
+            {
+                deltaMatrixScale.Scale(gContext.mScale);
+                memcpy(deltaMatrix, deltaMatrixScale.m16, sizeof(float) * 16);
+            }
+
+            if (!io.MouseDown[0])
+                gContext.mbUsing = false;
+
+            type = gContext.mCurrentOperation;
+        }
+    }
+
+    static void HandleRotation(float *matrix, float *deltaMatrix, int& type, float *snap)
+    {
+        ImGuiIO& io = ImGui::GetIO();
+        bool applyRotationLocaly = gContext.mMode == LOCAL;
+
+        if (!gContext.mbUsing)
+        {
+            type = GetRotateType();
+
+            if (type == ROTATE_SCREEN)
+            {
+                applyRotationLocaly = true;
+            }
+
+            if (io.MouseDown[0] && type != NONE)
+            {
+                gContext.mbUsing = true;
+                gContext.mCurrentOperation = type;
+                const vec_t rotatePlanNormal[] = { gContext.mModel.v.right, gContext.mModel.v.up, gContext.mModel.v.dir, -gContext.mCameraDir };
+                // pickup plan
+                if (applyRotationLocaly)
+                {
+                    gContext.mTranslationPlan = BuildPlan(gContext.mModel.v.position, rotatePlanNormal[type - ROTATE_X]);
+                }
+                else
+                {
+                    gContext.mTranslationPlan = BuildPlan(gContext.mModelSource.v.position, directionUnary[type - ROTATE_X]);
+                }
+
+                const float len = IntersectRayPlane(gContext.mRayOrigin, gContext.mRayVector, gContext.mTranslationPlan);
+                vec_t localPos = gContext.mRayOrigin + gContext.mRayVector * len - gContext.mModel.v.position;
+                gContext.mRotationVectorSource = Normalized(localPos);
+                gContext.mRotationAngleOrigin = ComputeAngleOnPlan();
+            }
+        }
+
+        // rotation
+        if (gContext.mbUsing)
+        {
+            gContext.mRotationAngle = ComputeAngleOnPlan();
+            if (snap)
+            {
+                float snapInRadian = snap[0] * DEG2RAD;
+                ComputeSnap(&gContext.mRotationAngle, snapInRadian);
+            }
+            vec_t rotationAxisLocalSpace;
+
+            rotationAxisLocalSpace.TransformVector(makeVect(gContext.mTranslationPlan.x, gContext.m
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') The function 'DrawCube' is never used. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)